### PR TITLE
Fix logger usage check for Log4j 2

### LINKED ISSUE #20243: Update logger usage check for Log4j 2
The logger usage check was designed for the ESLogger API, which was a simplified wrapper around the Log4j API. This check is useful, but the ESLogger wrapper will be removed after #20235 instead exposing the Log4j 2 API directly. The Log4j 2 API is significantly more complicated and the logger usage check will need to be updated for this API.

### PR DESCRIPTION
**Pull request by ywelsch**
With the switch to Log4j 2 the logger usage checker was temporarily disabled. This PR adapts the checks to work with Log4j 2 and re-enables the Gradle checks. It also fixes the wrong logger usages that have sneaked into the code base.

Closes #20243

### CODE REVIEW

**Comment by dakrone** [comment]
> It also fixes the wrong logger usages that have sneaked into the code base

<3

**Comment by ywelsch** [comment]
ping @jasontedor

File: test/logger-usage/src/main/java/org/elasticsearch/test/loggerusage/ESLoggerUsageChecker.java [hunk 1]
```diff
@@ -240,53 +251,141 @@ public void findBadLoggerUsages(MethodNode methodNode) {
...
+                        }
+
+                        int lengthWithoutMarker = argumentTypes.length - markerOffset;
+
+                        if (lengthWithoutMarker == 2 &&
+                            argumentTypes[markerOffset + 0].equals(STRING_CLASS) &&
+                            (argumentTypes[markerOffset + 1].equals(OBJECT_ARRAY_CLASS) ||
+                                argumentTypes[markerOffset + 1].equals(SUPPLIER_ARRAY_CLASS))) {
+                            // VARARGS METHOD: debug(Marker?, String, (Object...|Supplier...))
+                            checkArrayArgs(methodNode, logMessageFrames[i], arraySizeFrames[i], lineNumber, methodInsn, markerOffset + 0,
```
Comment (jasontedor) [inline test/logger-usage/src/main/java/org/elasticsearch/test/loggerusage/ESLoggerUsageChecker.java]: Can you explain why you wrote `markerOffset + 0`?

File: test/logger-usage/src/main/java/org/elasticsearch/test/loggerusage/ESLoggerUsageChecker.java [hunk 2]
```diff
@@ -240,53 +251,141 @@ public void findBadLoggerUsages(MethodNode methodNode) {
...
+                                markerOffset + 1);
+                        } else if (lengthWithoutMarker >= 2 &&
+                            argumentTypes[markerOffset + 0].equals(STRING_CLASS) &&
+                            argumentTypes[markerOffset + 1].equals(OBJECT_CLASS)) {
+                            // MULTI-PARAM METHOD: debug(Marker?, String, Object p0, ...)
+                            checkFixedArityArgs(methodNode, logMessageFrames[i], lineNumber, methodInsn, markerOffset + 0,
```
Comment (jasontedor) [inline test/logger-usage/src/main/java/org/elasticsearch/test/loggerusage/ESLoggerUsageChecker.java]: Same question here about `markerOffset + 0`?

File: test/logger-usage/src/main/java/org/elasticsearch/test/loggerusage/ESLoggerUsageChecker.java [hunk 3]
```diff
@@ -240,53 +251,141 @@ public void findBadLoggerUsages(MethodNode methodNode) {
...
+                                lengthWithoutMarker - 1);
+                        } else if ((lengthWithoutMarker == 1 || lengthWithoutMarker == 2) &&
+                            lengthWithoutMarker == 2 ? argumentTypes[markerOffset + 1].equals(THROWABLE_CLASS) : true) {
+                            // all the rest: debug(Marker?, (Message|MessageSupplier|CharSequence|Object|String|Supplier), Throwable?)
+                            checkFixedArityArgs(methodNode, logMessageFrames[i], lineNumber, methodInsn, markerOffset + 0, 0);
```
Comment (jasontedor) [inline test/logger-usage/src/main/java/org/elasticsearch/test/loggerusage/ESLoggerUsageChecker.java]: Same question here about `markerOffset + 0`?

**Review by jasontedor — COMMENTED** [review]
It looks great, thanks for fixing this, and it's surprisingly easy to read. I left one question.

File: test/logger-usage/src/main/java/org/elasticsearch/test/loggerusage/ESLoggerUsageChecker.java (hunk 1 above)
Comment (ywelsch) [inline test/logger-usage/src/main/java/org/elasticsearch/test/loggerusage/ESLoggerUsageChecker.java]: I wanted it to be more visual that this is the "first argument (= 0)" when ignoring the marker. I can remove it if you like, I don't have a strong preference on this.

Comment (jasontedor) [inline test/logger-usage/src/main/java/org/elasticsearch/test/loggerusage/ESLoggerUsageChecker.java]: I'm fine with it, I just wanted to understand why it was included.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by jasontedor** [comment]
Thanks @ywelsch, I'm good with the change assuming that you clean up the merge conflicts and the build passes.

**Pull request merged**